### PR TITLE
Select package to deploy (close #1917)

### LINKED ISSUE #1917: Fabric 2: Deploy View: When the view opens i can choose which package i want to deploy
image

### PR DESCRIPTION
**Pull request by Jakeeyturner**
Signed-off-by: Jake Turner <jaketurner25@live.com>

### CODE REVIEW

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepThree/DeployStepThree.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import IPackageRegistryEntry from '../../../../interfaces/IPackageRegistryEntry';
+
+interface IProps {
+    selectedPackage: IPackageRegistryEntry;
+    channelName: string;
+}
+
+class DeployStepThree extends Component<IProps> {
+
+    constructor(props: Readonly<IProps>) {
```
Comment (erin-hughes) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepThree/DeployStepThree.tsx:12]: You've a useless constructor in here

Comment (erin-hughes) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepThree/DeployStepThree.tsx:12]: You've a useless constructor here

File: packages/blockchain-ui/src/components/elements/ButtonList/ButtonList.tsx [hunk 2]
```diff
@@ -1,33 +1,53 @@
 import React, { Component } from 'react';
 import './ButtonList.scss';
 import { Button } from 'carbon-components-react';
+import IPackageRegistryEntry from '../../../interfaces/IPackageRegistryEntry';
 
-interface IProps {
-    buttons: any[];
+interface ButtonListProps {
+    onProgressChange: any;
```
Comment (erin-hughes) [inline packages/blockchain-ui/src/components/elements/ButtonList/ButtonList.tsx:7]: You can give your callback props types, for this one I think it'd be: 
`onProgressChange: (indexValue: number) => void;`

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepOne/DeployStepOne.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
+import React, { Component } from 'react';
+import { Dropdown, Accordion, AccordionItem } from 'carbon-components-react';
+import IPackageRegistryEntry from '../../../../interfaces/IPackageRegistryEntry';
+
+interface IProps {
+    packageEntries: IPackageRegistryEntry[];
+    selectedPackage: IPackageRegistryEntry | undefined; // Package the user has selected in dropdown
+    onPackageChange: any; // Callback to DeployPage
```
Comment (erin-hughes) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepOne/DeployStepOne.tsx:8]: You could give proper types to this callback prop, see previous comment

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
+import React, { Component } from 'react';
+import { TextInput, FileUploader, Accordion, AccordionItem, Link } from 'carbon-components-react';
+import IPackageRegistryEntry from '../../../../interfaces/IPackageRegistryEntry';
+
+interface IProps {
+    selectedPackage: IPackageRegistryEntry;
+    onDefinitionNameChange: any;
```
Comment (erin-hughes) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx:7]: You could give proper types to these callback props, see previous comment

File: packages/blockchain-ui/enzyme/tests/ButtonList.test.tsx [hunk 5]
```diff
@@ -1,24 +1,112 @@
-import React from 'react';
+import React, { ReactInstance, Component } from 'react';
```
Comment (erin-hughes) [inline packages/blockchain-ui/enzyme/tests/ButtonList.test.tsx:1]: Do you need `ReactInstance` and `Component` in this test file?

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepOne/DeployStepOne.tsx [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+        this.formatPackageEntries = this.formatPackageEntries.bind(this);
+        this.selectPackage = this.selectPackage.bind(this);
+
+    }
+
+    selectPackage(data: any): void {
+        const packageName: string = data.selectedItem.split('@')[0]; // Get everything before '@<version>'
+
+        const selectedPackage: IPackageRegistryEntry | undefined = this.state.packageEntries.find((_entry: IPackageRegistryEntry) => {
```
Comment (cazfletch) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepOne/DeployStepOne.tsx:31]: What if you had 2 packages with the same name but different versions?

File: packages/blockchain-ui/enzyme/tests/DeploySteps/DeployStepThree.test.tsx [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import renderer from 'react-test-renderer';
+import chai from 'chai';
+import sinon from 'sinon';
+import sinonChai from 'sinon-chai';
+import { ReactWrapper, mount } from 'enzyme';
```
Comment (erin-hughes) [inline packages/blockchain-ui/enzyme/tests/DeploySteps/DeployStepThree.test.tsx:6]: Not using this mount import

File: packages/blockchain-ui/src/components/elements/ButtonList/ButtonList.tsx [hunk 8]
```diff
@@ -1,33 +1,53 @@
...
+    currentIndex: number;
+    selectedPackage: IPackageRegistryEntry | undefined;
```
Comment (cazfletch) [inline packages/blockchain-ui/src/components/elements/ButtonList/ButtonList.tsx:9]: this feels wrong that you are passing data through the button list.

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+                                size='default'
+                            />
+                        </div>
+                    </div>
+
+                </div>
+                <div className='bx--row'>
+                    <div className='bx--col-lg-10'>
+                        <Accordion>
+                            <AccordionItem title={'Learn about other smart contract definition parameters (advanced)'}>
```
Comment (cazfletch) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx:86]: When you trigger the accordion the next button moves

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+                        The package's name and version (where available) are copied across to the definition as defaults.
+                        You may optionally change them.
+                    </div>
+                </div>
+                <div className='bx--row'>
+                    <h5>Smart contract definition</h5>
+                </div>
+                <div className='bx--row margin-bottom-06'>
+                    <div className='bx--col-lg-5 margin-right-06'>
+                        <TextInput id='nameInput' labelText='Definition name' defaultValue={this.props.currentDefinitionName ? this.props.currentDefinitionName : this.props.selectedPackage.name} onChange={this.handleDefinitionNameChange}></TextInput>
```
Comment (cazfletch) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx:45]: Can you add an issue for error handling

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+                    </div>
+                    <div className='bx--col-lg-5'>
+                        <TextInput id='versionInput' labelText='Definition version' defaultValue={this.props.currentDefinitionVersion ? this.props.currentDefinitionVersion : this.props.selectedPackage.version} onChange={this.handleDefinitionVersionChange}></TextInput>
```
Comment (cazfletch) [inline packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx:48]: next button isn't disabled if you don't have a value for name or version

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx (hunk 10 above)
Comment (Jakeeyturner) [reply packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx:45]: Added to https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues/1922

File: packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx (hunk 11 above)
Comment (Jakeeyturner) [reply packages/blockchain-ui/src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo.tsx:48]: Added to https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues/1922

File: packages/blockchain-ui/src/components/elements/ButtonList/ButtonList.tsx (hunk 8 above)
Comment (Jakeeyturner) [reply packages/blockchain-ui/src/components/elements/ButtonList/ButtonList.tsx:9]: Fixed - we now determine whether the Next button should be disabled in the DeployPage.

**Pull request merged**